### PR TITLE
Render notice about payment method

### PR DESCRIPTION
**Pull request by manzurahammed**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Display notice when there are expiring MP subscriptions.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Local WCCOM

Checkout branch `update/helper-notice-missing-payment-method` for updated endpoint response.

#### Local WC

**Missing payment method and one expiring subscription**

* Make sure to to remove payment method of connected account on local wccom
* Has one expiring subscription
* Check WC Setting and Extension pages.
* See screenshot below

image

image

**Missing payment method and more than one expiring subscriptions**

Similar to above, but plural on the messaging:

> Your WooCommerce extension subscriptions are missing...

**Payment method is available and has expiring subscriptions**

Should be the same experience as described in https://github.com/woocommerce/woocommerce/pull/47004

**CTA**

Add payment method button and link go to https://woocommerce.com/my-account/add-payment-method/

**Tracks**

* Viewing notice in WC settings page: `wcadmin_woo_subscription_expiring_notice_in_settings_shown`
* Viewing notice in Extension page: `wcadmin_woo_subscription_expiring_notice_in_marketplace_shown`
* Dismissing notice in WC setting pages: `wcadmin_woo_subscription_expiring_notice_in_settings_dismissed`
* Dismissing notice in Extension page: `wcadmin_woo_subscription_expiring_notice_in_marketplace_dismissed`
* Add payment method in WC setting pages: `wcadmin_woo_subscription_expiring_notice_in_settings_clicked`
* Add payment method in Extension page: `wcadmin_woo_subscription_expiring_notice_in_marketplace_clicked`

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php [hunk 1]
```diff
@@ -2261,6 +2261,43 @@ public static function get_install_base_url() {
...
+			return array();
+		}
+
+		$data = json_decode( wp_remote_retrieve_body( $request ), true );
+
+		if ( empty( $data ) || ! is_array( $data ) ) {
+			$data = array();
+		}
+
+		set_transient( $cache_key, $data, 12 * HOUR_IN_SECONDS );
```
Comment (gedex) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper.php]: This feels too long, `/subscriptions` is cached for 1h.

File: plugins/woocommerce/src/Admin/PluginsHelper.php [hunk 2]
```diff
@@ -821,6 +821,35 @@ function ( $sub ) {
 
 		$total_expiring_subscriptions = count( $expiring_subscriptions );
 
+		// When payment method is missing on WooCommerce.com.
+		$helper_notices = WC_Helper::get_notices();
```
Comment (hesahesa) [inline plugins/woocommerce/src/Admin/PluginsHelper.php:825]: this notice will be keyed (for dismissal / reshow purpose) with the same key as the expiring subscription notices key. Also this notice will use the same Track events as expiring subscription notices. Is this intentional?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php (hunk 1 above)
Comment (hesahesa) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper.php]: Once a customer added a payment method in WCCOM, we will still show the payment method not available notice for up to 12 hours in this case. IMHO we should put this cache duration shorter

File: plugins/woocommerce/src/Admin/PluginsHelper.php (hunk 2 above)
Comment (gedex) [reply plugins/woocommerce/src/Admin/PluginsHelper.php:825]: Yes, it's intentional as we'd like to show at most 1 notice and payment method only relevant when the subscription is expiring. The priority of what notice gets rendered will be:

1. Expired
2. Expiring and missing payment method
3. Expiring

**Review by hesahesa — APPROVED** [review]
LGTM, there is a minor concern about the transient duration

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php (hunk 1 above)
Comment (gedex) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper.php]: Updated to 1h in https://github.com/woocommerce/woocommerce/pull/47141/commits/d4190c98ed8929626ec0f04169d31170f4f36e47

**Pull request merged**